administrator/components/com_associations/models/fields/associatedcomponent.php

```diff
@@ -48,29 +73,30 @@ protected function getGroups()
 			$componentModelsPath = $componentAdminPath . '/models';
 
 			// Only components that exist also in the site client, aren't in the excluded components array and have models.
-			if (!is_dir(JPATH_SITE . '/components/' . $component) || !is_dir($componentModelsPath) || in_array($component, $excludeComponents))
+			if (!is_dir($componentModelsPath) || in_array($component, $excludeComponents))
 			{
 				continue;
 			}
 
 			// Check if component uses associations, by checking is models.
 			foreach (glob($componentModelsPath . '/*.php', GLOB_NOSORT) as $modelFile)
 			{
-				$file = file_get_contents($modelFile);
+				$itemName = ucfirst(basename($modelFile, '.php'));
 
-				// Check if this model uses associations. Add component model option to select box if so.
-				if (strpos($file, 'protected $associationsContext'))
-				{
-					$modelNameSpace = ucfirst(basename($modelFile, '.php'));
-
-					JModelLegacy::addIncludePath($componentModelsPath, $modelNameSpace . 'Model');
-					$model = JModelLegacy::getInstance($modelNameSpace, $componentName . 'Model', array('ignore_request' => true));
+				JLoader::register($componentName . 'Model' . $itemName, $modelFile);
+				$model = JModelLegacy::getInstance($itemName, $componentName . 'Model', array('ignore_request' => true));
 
+				// Check if this model uses associations.
+				if ($model && $model->get('associationsContext') && $model->get('typeAlias') && !in_array($model->get('typeAlias'), $typeAliasList))
+				{
 					// Load component language file.
+					$lang->load($component . '.sys', JPATH_ADMINISTRATOR) || $lang->load($component . '.sys', $componentAdminPath);
 					$lang->load($component, JPATH_ADMINISTRATOR) || $lang->load($component, $componentAdminPath);
 
-					// Add componet option select box.
-					$options[JText::_($component)][] = JHtml::_('select.option', $model->typeAlias, JText::_($component));
+					// Add component option select box.
+					$options[JText::_($component)][] = JHtml::_('select.option', $model->get('typeAlias'), JText::_($component));
```

Comment:
for flexility with components with several models we need to have a way to use a correct text variable that itentifies the item inside the component

Proposal:
```php
$options[JText::_($component)][] = JHtml::_('select.option', $model->get('typeAlias'), JText::_($component));
```

Should be
```php
$options[JText::_($component)][] = JHtml::_('select.option', $model->get('typeAlias'), JText::_($component . '_SUBMENU_' . $itemName));
```